### DISCUSSION TITLE
Python: what's the right `poetry` and `uv` dependency management workflow?

### DISCUSSION THREAD

**Discussion by phitoduck** [projen/projen#3875] [Q&A]
Hi there, I just created a trivial Python `poetry` project with `projen`. 

I know that `projen` wants to lock down `pyproject.toml`, but  `poetry add <pkg>` and `uv add <pkg>` each modify `pyproject.toml` directly.

Are `projen` Poetry users simply not meant to run the `uv add ...` and `poetry add ...` commands, and instead add dependencies using the usual `addDependency(...)` or `deps=[...]` abstractions?

Does this get in the way of anything? Do the JS/TS people have to do the same thing given that `package.json` is locked? (basically, avoid `pnpm add ...` and its equivalents)

I ask, because I'm considering writing a set of projen components for a new team that uses `uv`. `uv` is really hot right now and I'm worried, they'll revolt if the dependency management experience isn't "uv-native" enough.

If there were a way to use `uv add` as intended, while still having the ability to push changes to other areas in the `pyproject.toml` using `projen`--that'd be so amazing.

Here's my setup:

```python
from projen.python import PythonProject

project = PythonProject(
    author_email="xxx",
    author_name="phitoduck",
    module_name="projen_poetry",
    name="projen-poetry",
    version="0.1.0",
    poetry=True,
)

project.synth()
```

<img width="1178" alt="image" src="https://github.com/user-attachments/assets/3eaccb1c-d707-4f94-bdfa-00f21ae1b90a">

I wish there were a way to make the `project.dependencies` portion of the toml file writable, and lock down the rest of the file, lol.

**Comment by mrgrain** [comment]
Yes, in node projects the user is not expected to run `npm install ...` but instead modify the .projenrc file and then run projen which takes care of the rest. There's been plenty of discussion around it. The short answer is that other approaches might be possible, but that's the only one that currently exists.

For node, the `package.json` is actually already under shared management. Specifically the versions of "unmanaged" dependencies are read and accepted as state by projen. There's no reason a component couldn't extended this ideas to dependencies themselves.

You will likely have to build a custom python project type from scratch (with inspiration from what exists), as well as some components that don't do today what you want them to.
